File: leetcode/2024_8_august/python/11_minimum_number_of_days_to_disconnect_island.py
# CODE

import copy
from typing import List

class Solution:
    def solve(self, i, j, m, n, grid):
        if grid[i][j] == 0:
            return 
        grid[i][j] = 0
        if i - 1 >= 0:
            self.solve(i - 1, j, m, n, grid)
        if i + 1 < m:
            self.solve(i + 1, j, m, n, grid)
        if j - 1 >= 0:
            self.solve(i, j - 1, m, n, grid)
        if j + 1 < n:
            self.solve(i, j + 1, m, n, grid)


    def island_count(self, grid):
        cnt = 0
        m, n = len(grid), len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    cnt += 1
                    self.solve(i, j, m, n, grid)
        return cnt

    def minDays(self, grid: List[List[int]]) -> int:
        days = 0
        new_grid = copy.deepcopy(grid)
        cnt = self.island_count(new_grid)
        if cnt != 1:
            return days
        
        days = 1
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                new_grid = copy.deepcopy(grid)
                new_grid[i][j] = 0
                cnt = self.island_count(new_grid)
                if cnt != 1:
                    return days
        
        days = 2
        return days

File: leetcode/2024_8_august/c++/11_minimum_number_of_days_to_disconnect_island.cpp
// TIME COMPEXITY
// O((n * m) ^ 2)

// SPACE COMPEXITY
// O(n * m)

// CODE

class Solution {
public:
    int n,m;
    void dfs(vector<vector<int>>& grid, int i, int j, vector<vector<bool> > &vis){
        if(i < 0 || i >= n || j < 0 || j >= m || vis[i][j] || grid[i][j] == 0) return ;

        vis[i][j] = true;
        dfs(grid,i+1,j,vis);
        dfs(grid,i-1,j,vis);
        dfs(grid,i,j+1,vis);
        dfs(grid,i,j-1,vis);
    }
    int countf(vector<vector<int>>& grid){
        int cnt = 0;
        vector<vector<bool> > vis(n,vector<bool>(m));

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 1 && !vis[i][j]){
                    dfs(grid,i,j,vis);
                    cnt++;
                }
            }
        }
        return cnt;
    }
    int minDays(vector<vector<int>>& grid) {
        n = grid.size();
        m = grid[0].size();

        int cnt = countf(grid);

        if(cnt > 1 || cnt == 0) return 0;
        else{
            for(int i = 0; i < n; i++){
                for(int j = 0; j < m; j++){
                    if(grid[i][j] == 1){
                        grid[i][j] = 0;
                        cnt = countf(grid);
                        if(cnt > 1 || cnt == 0) return 1;
                        grid[i][j] = 1;
                    }
                }
            }
        }
        return 2;
    }
};